design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,241 @@
+---
+title: swift-and-keystone
+target-version: release-X.X
+---
+
+# Swift and Keystone Integration
+
+## Summary
+
+### Goals
+
+The goal of this proposal is to allow configuring the Swift API and
+the Keystone integration of Ceph RGW natively via the Object Store
+CRD, which allows native integration of the Rook operated Ceph RGW into
+OpenStack Clouds.
+
+Both changes are bundled together as one proposal because they will
+typically deployed together. It is unlikely to encounter a Swift
+object store API outside an OpenStack cloud and in the context of an
+OpenStack cloud the Swift API needs to be integrated with the Keystone
+authentication service.
+
+Any changes to the CRD must be future safe and cleanly allow extension
+to further technologies (such as LDAP authentication).
+
+### Non-Goals
+
+* Support for Keystone API versions below v3. API version v2 has long
+  been deprecated [has been removed in
+  Queens](https://docs.openstack.org/keystone/latest/contributor/http-api.html)
+  which was release in 2018 and is now in extended maintenance mode
+  (which means it gets no more points releases and only sporadic bug
+  fixes/security fixes).
+
+* Authenticating Ceph RGW to Keystone via admin token – Only
+  authentication via an OpenStack service account will be supported.
+
+* Support for APIs beside S3 and Swift.
+
+* Interaction of Swift with OBCs is out of scope of this document.  If
+  you need to access a bucket created by OBC via Swift you need to
+  create a separate `cephobjectstoreuser`, configure its access rights
+  to the bucket and use those credentials.
+
+* Support for authentication technologies other than Keystone (e.g. LDAP)
+
+* Exposing options that disable security features (e.g. TLS verification)
+
+## Proposal details
+
+The Object Store CRD will have to be extended to accommodate the new
+settings.
+
+### Keystone integration
+
+A new section `auth:` is added to the Object Store CRD. To configure
+the keystone integration:
+
+```yaml
+auth:
+  keystone:
+    url: https://keystone:5000/
+    acceptedRoles: ["_member_", "service", "admin"]
+    apiVersion: 3
+    implicitTenants: swift
+    tokenCacheSize: 1000
+    revocationInterval: 1200
+    serviceUserSecret: rgw-service-user
+```
+
+The certificate to verify the Keystone endpoint can't be explicitly
+configured in Ceph RGW, the system configuration of the pod is used,
+you can add to the system certificate store via the
+`gateway.caBundleRef` setting of the object store CRD.
+
+The credentials for the Keystone service account used by Ceph RGW are
+supplied in a Secret that contains a mapping of OpenStack openrc
+environment variables. Only password authentication to Keystone is
+supported. Example:
```

Comment:
> Only password authentication to Keystone is supported.

Does this mean that we **choose** to only support password auth, or is that a limitation of keystone?

Reply:
It is a limitation of RGW, I've made it explicit in the newest version.